### PR TITLE
[CALCITE-6735] Type coercion for comparisons does not coerce ROW types

### PR DESCRIPTION
**Pull request by mihaibudiu**
I also had to implement in the RexToLixTranslator casts between ROW types.

### CODE REVIEW

**Comment by mihaibudiu** [comment]
The CI seems finished, just not detected properly by the UI.

**Comment by mihaibudiu** [comment]
In practical terms, what this PR does is to convert a comparison of ROW values with different types to a common type using casts.

File: core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java [hunk 1]
```diff
@@ -629,6 +630,28 @@ private RelDataType getTightestCommonTypeOrThrow(
...
+      if (typeName2 != typeName1) {
+        return null;
+      }
+      if (type1.getFieldCount() != type2.getFieldCount()) {
+        return null;
+      }
+      List<RelDataTypeField> leftFields = type1.getFieldList();
+      List<RelDataTypeField> rightFields = type2.getFieldList();
+      List<RelDataTypeField> resultFields = new ArrayList<>();
+      for (int i = 0; i < leftFields.size(); i++) {
```
Comment (NobiGo) [inline core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java:643]: Need to confirm the list size is equal?

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java:643]: That is done a few lines above

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java:681]: @NobiGo that's line 675, comparing getFieldCount().

**Review by NobiGo — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java [hunk 2]
```diff
@@ -331,6 +332,35 @@ private Expression getConvertExpression(
       return Expressions.convert_(cast, typeFactory.getJavaClass(nullableTarget));
     }
 
+    if (targetType.getSqlTypeName() == SqlTypeName.ROW) {
+      assert sourceType.getSqlTypeName() == SqlTypeName.ROW;
+      List<RelDataTypeField> targetTypes = targetType.getFieldList();
+      List<RelDataTypeField> sourceTypes = sourceType.getFieldList();
+      assert targetTypes.size() == sourceTypes.size();
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:339]: Why do we need to judge this? The sizes of targettype and sourcetype may be different?

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:339]: The validator will reject such programs.

**Pull request merged**